Handle default body text when creating issues and pull requests

chemotaxis opened the pull request:
This comment refers to the original description: https://github.com/cli/cli/pull/3658#pullrequestreview-663009869

This fixes #674.

After a second look at the code, `surveyext.GhEditor` explicitly returns an empty string when pressing `Enter`.  It doesn't look like it had anything to do with unwritten temporary files. I added an extra if clause to see if the editor prompt was instructed to add the default value to the editor.

Unfortunately, adding tests for this isn't straightforward and, so far, no tests exist for `survey.GhEditor`.  If you still want me to try to add tests, let me know.

The `GhEditor` is used in these cases:

- `gh issue create`
- `gh issue edit`
- `gh pr create`
- `gh pr edit`
- `gh pr review`

`gh issue create` and `gh pr create` are the only places where the new condition will trigger.  The original behavior (returning an empty string when skipping) seem to be preserved everywhere else.

mislav reviewed (changes requested):
> I wasn't able to fully trace why the commit messages were erased when skipping during `gh pr create`. However, my best guess is that, because the editor is never opened, the default text is never written to a temporary file. Because nothing was written, the survey returns nothing and overwrites the commit messages in `state.Body` with an empty string.

That sounds about right! Thanks so much for taking a deep dive into this.

I do, however, feel somewhat strongly about fixing this _without_ changing the actual Survey prompt around the editor for now. The prompt could definitely use some redesign, but I would want that redesign to be motivated by enhancing usability instead of trying to fix a bug. In other words, I would like to see the change to rethink the prompt be separate from the change to fix this bug, and when we do change the prompt, we would have to consider whether it's backward-compatible with users for whom the old prompt (press `e` to open editor) is already part of muscle memory.

Do you think it's possible to keep using GHEditor and just handle the case of preserving the body if the editor wasn't invoked?

chemotaxis committed:
Return default text if skipping the text editor when prompted

If we are allowed to skip the editor _and_ we want to append the default
text to the editor if we'd opened it, we just return the default text.

Co-Authored-By: Mislav Marohnić <mislav@github.com>

chemotaxis commented on pkg/surveyext/editor.go:
```diff
@@ -104,7 +104,9 @@ func (e *GhEditor) prompt(initialValue string, config *survey.PromptConfig) (int
 			break
 		}
 		if r == '\r' || r == '\n' {
-			if e.BlankAllowed {
```
Let me know if you'd like a nested if statement or just keep the flat if statement.

mislav replied:
You are welcome to restructure the conditional logic to something that would be more cleaner in your view.

chemotaxis replied:
This is resolved in commit 91d4ae117d9704f1dc675c10222b722eab18b662.

chemotaxis commented on pkg/surveyext/editor.go:
```diff
@@ -104,7 +104,9 @@ func (e *GhEditor) prompt(initialValue string, config *survey.PromptConfig) (int
 		}
 		if r == '\r' || r == '\n' {
-			if e.BlankAllowed {
+			if e.BlankAllowed && e.AppendDefault {
```
It also might make sense to rename `BlankAllowed` to something like `SkipEditorAllowed`, since skipping no longer just returns an empty string.

chemotaxis commented on pkg/surveyext/editor.go:
```diff
@@ -104,7 +104,9 @@ func (e *GhEditor) prompt(initialValue string, config *survey.PromptConfig) (int
 			break
 		}
 		if r == '\r' || r == '\n' {
-			if e.BlankAllowed {
```
It also might make sense to rename `BlankAllowed` to something like `SkipEditorAllowed`, since skipping no longer just returns an empty string.

mislav replied:
Makes sense to me too! Thanks for looking into this

chemotaxis replied:
Sure!  This is taken care of in commit 6497114a864da1ddb9a52031d67d7bbba523bf10.

chemotaxis commented:
I think this fix is done.  The one additional thing I did was convert the list of if statements for detecting keyboard presses to a switch statement (commit 6db87ff56f199fd8022d80154d9ed38687a099ff).

chemotaxis reviewed (commented):
I reverted the initial changes and have a solution at commit 6db87ff56f199fd8022d80154d9ed38687a099ff.

chemotaxis commented:
Hi @mislav!  I'm just checking to see if there is something else I need to fix in this pull request.  It looks like `pr-auto` is failing because of how I edited my pull request description, I think?

mislav committed:
Add test for our survey editor extension

mislav reviewed (approved):
Thank you for your patience! The change looks good, but there was a lot of renaming going on, and even though I agree with your motivations for renaming, right now I wanted us to just focus on making the smallest change possible to fix the bug. I've condensed the fix even further and added a test.

chemotaxis commented:
Understood.  Thanks for adding the tests and for merging!

Do those renamings or refactorings manifest as separate issues that I can contribute to or do those kinds of issues get taken care of periodically by the core team?

mislav commented:
@chemotaxis We periodically address technical debt by refactoring and shuffling code around, of course! We are open to receiving those kinds of contributions as well, but keep in mind that we expect that a refactoring yields some tangible improvement (e.g. that the new code is slimmer, or easier to test, or more portable, or more resilient against edge cases) and that the author is mindful of potential regressions (e.g. avoids changes to code that has little test coverage). Otherwise, if a change is purely cosmetic, such as renaming a struct field (even if the new name is slightly more descriptive), we are less keen on accepting those since the risk-to-benefit ratio isn't in our favor.
